active/0024-message-validation.md

```diff
@@ -0,0 +1,132 @@
+# Message Filter and Validation
+
+## Changelog
+
+* 2023-10-31: @zmstone Initial draft message validation, filter and transformation.
+* 2022-12-19: @zmstone Limit the scope to filter and validation.
+
+## Abstract
+
+A new feature for EMQX v5 to validate messages.
+
+## Motivation
+
+### Message Validation
+
+Technically, by using EMQX rules engine, it is possible to validate the incoming
+MQTT messages before sending the message off to data bridges.
+
+For instance, one can write a rules engine SQL like below to
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but they do not interfere MQTT message pub/sub.
+For example, if there is client subscribed to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
+
+### Message filter
+
+Message filtering can be achieved by confugring message validation rule to drop invalid messages.
+
+## Design
+
+This data validation step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Make use of the rules engine SQL, but without the `FROM` clause as it's always the MQTT message being th input.
+
+Each data validation rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+### Message Validation
+
+```
+validation1 {
+    enable = true # or 'false'
+    type = validation
+    description = "drop message if it is not compabitble with my avro schema or if payload.value is less tha 0"
+    topics = "t/#" # or topics = ["t/1", "t/2"]
+    check_mode = any_pass # or all_pass
```

Comment:
Ben suggests renaming to
```
{
  validation_strategy = "and" | "or" | "AND" | "OR"
}
```

Reply:
I do not have strong opinion on the name of the config, but would strongly recommend `any` `all` instead of `and` `or`.

Reply:
I believe "and" | "or" is more akin to logical expressions in code, making it more familiar to developers. What are the advantages of "all" | "any", and are there any other products that use it?

Reply:
not just `any` and `all`. but `any_pass` and `all_pass` 
meaning is clear to developers an non-developers too.

Reply:
You've convinced me, now I'll go convince Ben.